source/agora/script/ScopeCondition.d

```diff
@@ -0,0 +1,229 @@
+/*******************************************************************************
+
+    Keeps track of scopes and their conditions (TRUE or FALSE).
+    This struct can be used to implement conditional (IF/ELSE/ENDIF) logic.
+
+    It does this pushing a new scope for each visited `IF` opcode,
+    popping a scope for every visited `ENDIF` opcode, and toggling the scope's
+    condition for every visited `ELSE` opcode.
+
+    Unlike C-like programming languages, we do not support GOTO and therefore
+    may only increment the program counter one instruction at a time.
+
+    This implementation is largely based on Bitcoin's `ConditionStack`,
+    as it's the most optimal O(1) solution we can think of.
+
+    For a description on how code flow control works (for a previous version),
+    see: https://building-on-bitcoin.com/docs/slides/Thomas_Kerin_BoB_2018.pdf
+
+    Copyright:
+        Copyright (c) 2009-2010 Satoshi Nakamoto
+        Copyright (c) 2009-2020 The Bitcoin Core developers
+        Copyright (c) 2020 BOS Platform Foundation Korea
```

Comment:
I'm not actually sure this passes the legalese. The code is based on Bitcoin's implementation, but I reordered some code and renamed a bunch of the fields. I did this to make the code clearer to understand.